v-next/hardhat-node-test-runner/src/task-action.ts

```diff
@@ -0,0 +1,54 @@
+import type { HardhatConfig } from "@ignored/hardhat-vnext/types/config";
+import type { NewTaskActionFunction } from "@ignored/hardhat-vnext/types/tasks";
+
+import { resolve } from "node:path";
+import { run } from "node:test";
+import { fileURLToPath } from "node:url";
+
+import hardhatTestReporter from "@ignored/hardhat-vnext-node-test-reporter";
+import { getAllFilesMatching } from "@ignored/hardhat-vnext-utils/fs";
+
+interface TestActionArguments {
+  testFiles: string[];
+  only: boolean;
+}
+
+function isTypescriptFile(path: string): boolean {
+  return /\.(ts|cts|mts)$/i.test(path);
+}
+
+function isJavascriptFile(path: string): boolean {
+  return /\.(js|cjs|mjs)$/i.test(path);
+}
+
+async function getTestFiles(
+  testFiles: string[],
+  config: HardhatConfig,
+): Promise<string[]> {
+  if (testFiles.length !== 0) {
+    const testFilesAbsolutePaths = testFiles.map((x) =>
+      resolve(process.cwd(), x),
+    );
+
+    return testFilesAbsolutePaths;
+  }
+
+  return getAllFilesMatching(
+    config.paths.tests,
+    (f) => isJavascriptFile(f) || isTypescriptFile(f),
+  );
+}
+
+const testWithHardhat: NewTaskActionFunction<TestActionArguments> = async (
+  { testFiles, only },
+  hre,
+) => {
+  const files = await getTestFiles(testFiles, hre.config);
+
+  const tsx = fileURLToPath(import.meta.resolve("tsx/esm"));
+  process.env.NODE_OPTIONS = `--import ${tsx}`;
+
+  run({ files, only }).compose(hardhatTestReporter).pipe(process.stdout);
```

Comment:
Don't we have to await this or something? How can we get to know if the tests failed?

We should set `process.exitCode` if it fails, and also return some error code.

Reply:
It's not a promise, it's a stream. Failed tests get piped through to `process.stdout`

Reply:
I went deep into this rabbit hole, and came out with this example snippet showing how to do these things

```ts
import path from "node:path";
import { finished } from "node:stream/promises";
import { run } from "node:test";
import { tap } from "node:test/reporters";

async function runTests(): Promise<number> {
  let failures = 0;
  // `files:` resolves relative paths, not globs, that's ok.
  const reporterStream = run({ files: [path.resolve("./test.ts")] })
    .on("test:fail", () => failures++)
    .compose(tap);

  reporterStream.pipe(process.stdout);

  // This should use the reporter's stream, not the one piped to stdout, as it
  // never finishes, as stdout is never closed.
  await finished(reporterStream);

  return failures;
}

const failures = await runTests();
console.log("Failures:", failures);

if (failures > 0) {
  process.exitCode = failures;
}
```

Here we:

1. Count the test failures and return it.
2. Set `process.exitCode` if needed.
3. Await for the test run to finish.